File: sdks/typescript/src/runtime/join.ts
export function joinBy<I, R, O>(
    inputs: I[],
    results: R[],
    inputKey: (input: I) => unknown,
    resultKey: (result: R) => unknown,
    combine: (result: R, input: I) => O,
): O[] {
    const byKey = new Map(inputs.map((input) => [inputKey(input), input]));

    return results.map((result) => {
        const input = byKey.get(resultKey(result));

        if (input === undefined) {
            throw new Error(`joinBy: no input matches key ${resultKey(result)}`);
        }

        return combine(result, input);
    });
}

File: sdks/python/src/pecans/runtime/join.py
from __future__ import annotations

from typing import Callable, TypeVar

TInput = TypeVar("TInput")
TResult = TypeVar("TResult")
TOutput = TypeVar("TOutput")


def join_by(
    inputs: list[TInput],
    results: list[TResult],
    input_key: Callable[[TInput], object],
    result_key: Callable[[TResult], object],
    combine: Callable[[TResult, TInput], TOutput],
) -> list[TOutput]:
    by_key = {input_key(item): item for item in inputs}

    def join(result: TResult) -> TOutput:
        item = by_key.get(result_key(result))

        if item is None:
            raise ValueError(f"join_by: no input matches key {result_key(result)}")

        return combine(result, item)

    return [join(result) for result in results]
